File: src/test/java/com/automation/um/pages/ProductsPage.java
package com.automation.um.pages;

import com.automation.um.utils.WaitUtils;
import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class ProductsPage extends BasePage {
  public ProductsPage(WebDriver driver) {
    super(driver);
  }

  public void openProducts() {
    open("/products");
    waitForGridReady();
  }

  public void waitForGridReady() {
    WaitUtils.invisible(driver, testId("product-card-skeleton"));
    if (driver.findElements(testId("products-grid")).isEmpty()) {
      wait.until(d -> d.getPageSource().contains("No products found"));
    } else {
      el("products-grid");
    }
  }

  public void search(String query) {
    fillInput("products-search", query);
    WaitUtils.debounce();
    waitForGridReady();
  }

  public void filterByStatus(String status) {
    selectOptionByValue("products-filter-status", status);
    waitForGridReady();
  }

  public void sortBy(String optionValue) {
    selectOptionByValue("products-sort", optionValue);
    waitForGridReady();
  }

  public void setPageSize(int size) {
    selectOptionByValue("products-page-size", String.valueOf(size));
    waitForGridReady();
  }

  public void refresh() {
    el("products-refresh").click();
    waitForGridReady();
  }

  public void toggleProduct(String productId) {
    el("product-toggle-" + productId).click();
    WaitUtils.invisible(driver, testId("products-saving"));
  }

  public void nextPage() {
    el("products-next").click();
    waitForGridReady();
  }

  public int visibleCardCount() {
    List<WebElement> cards = driver.findElements(By.cssSelector("[data-testid^='product-card-']"));
    return (int) cards.stream().filter(c -> !"product-card-skeleton".equals(c.getAttribute("data-testid"))).count();
  }

  public String firstProductId() {
    return driver.findElements(By.cssSelector("[data-testid^='product-card-']")).stream()
        .map(el -> el.getAttribute("data-testid"))
        .filter(id -> id != null && !id.equals("product-card-skeleton"))
        .map(id -> id.replace("product-card-", ""))
        .findFirst()
        .orElse(null);
  }

  public String selectedCountText() {
    return el("products-selected-count").getText();
  }

  public String productToggleText(String productId) {
    return el("product-toggle-" + productId).getText();
  }

  public String pageText() {
    return el("products-page").getText();
  }

  public boolean isNextEnabled() {
    return el("products-next").isEnabled();
  }
}

File: playwright/pages/ProductsPage.js
import { BasePage } from "./BasePage.js";
import * as WaitUtils from "../utils/waitUtils.js";

export class ProductsPage extends BasePage {
  async openProducts() {
    await this.open("/products");
    await this.waitForGridReady();
  }

  async waitForGridReady() {
    await WaitUtils.invisible(this.page, "product-card-skeleton").catch(() => {});
    const hasGrid = await this.testId("products-grid").count();
    if (hasGrid === 0) {
      await this.page.getByText("No products found").waitFor({ state: "visible" }).catch(() => {});
    } else {
      await this.el("products-grid");
    }
  }

  async search(query) {
    await this.fillInput("products-search", query);
    await WaitUtils.debounce();
    await this.waitForGridReady();
  }

  async filterByStatus(status) {
    await this.selectOptionByValue("products-filter-status", status);
    await this.waitForGridReady();
  }

  async sortBy(optionValue) {
    await this.selectOptionByValue("products-sort", optionValue);
    await this.waitForGridReady();
  }

  async setPageSize(size) {
    await this.selectOptionByValue("products-page-size", String(size));
    await this.waitForGridReady();
  }

  async refresh() {
    await this.testId("products-refresh").click();
    await this.waitForGridReady();
  }

  async toggleProduct(productId) {
    await this.testId(`product-toggle-${productId}`).click();
    await WaitUtils.invisible(this.page, "products-saving").catch(() => {});
  }

  async nextPage() {
    await this.testId("products-next").click();
    await this.waitForGridReady();
  }

  async visibleCardCount() {
    const cards = this.page.locator("[data-testid^='product-card-']");
    const count = await cards.count();
    let visible = 0;
    for (let i = 0; i < count; i++) {
      const id = await cards.nth(i).getAttribute("data-testid");
      if (id && id !== "product-card-skeleton") visible++;
    }
    return visible;
  }

  async firstProductId() {
    const cards = this.page.locator("[data-testid^='product-card-']");
    const count = await cards.count();
    for (let i = 0; i < count; i++) {
      const id = await cards.nth(i).getAttribute("data-testid");
      if (id && id !== "product-card-skeleton") {
        return id.replace("product-card-", "");
      }
    }
    return null;
  }

  async selectedCountText() {
    return (await this.testId("products-selected-count").textContent())?.trim() ?? "";
  }

  async productToggleText(productId) {
    return (await this.testId(`product-toggle-${productId}`).textContent())?.trim() ?? "";
  }

  async pageText() {
    return (await this.testId("products-page").textContent())?.trim() ?? "";
  }

  async isNextEnabled() {
    return this.testId("products-next").isEnabled();
  }
}
